controllers/tykoasapidefinition_controller.go

```diff
@@ -188,6 +220,173 @@ func (r *TykOasApiDefinitionReconciler) Reconcile(ctx context.Context, req ctrl.
+			return err
+		}
+	}
+
+	// this cleans up dangling versioned OAS APIs
+	if len(versionedTykOAS.Status.BaseVersionName) > 0 {
+		var baseOasAPI v1alpha1.TykOasApiDefinition
+
+		ns := versionedTykOAS.Status.BaseVersionNamespace
+		if ns == "" {
+			ns = "default"
+		}
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      versionedTykOAS.Status.BaseVersionName,
+			Namespace: ns,
+		}, &baseOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("base oas version not found, detaching version from the base OAS API.")
+
+				versionedTykOAS.RemoveOASVersionStatus()
+
+				if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+					return err
+				}
+
+				return nil
+			}
+
+			return err
+		}
+
+		var oasAPIPresent bool
+
+		for _, v := range baseOasAPI.Spec.Versioning.Versions {
+			if v.OasApiDefinitionRef == versionedTykOAS.Name {
+				oasAPIPresent = true
+			}
+		}
+
+		if !oasAPIPresent {
+			versionedTykOAS.RemoveOASVersionStatus()
+
+			if err := r.Status().Update(ctx, versionedTykOAS); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+
+	return nil
+}
+
+func (r *TykOasApiDefinitionReconciler) handleVersioningTykOASAPI(ctx context.Context,
+	tykBaseOASCrd *v1alpha1.TykOasApiDefinition,
+	log logr.Logger,
+	tykBaseOASDoc,
+	baseAPIID string,
+) error {
+	var versions []*TykOasApiDefinitionVersion
+
+	for _, v := range tykBaseOASCrd.Spec.Versioning.Versions {
+		var versionOasAPI v1alpha1.TykOasApiDefinition
+
+		if err := r.Get(ctx, types.NamespacedName{
+			Name:      v.OasApiDefinitionRef,
+			Namespace: v.Namespace,
+		}, &versionOasAPI); err != nil {
+			if k8serrors.IsNotFound(err) {
+				log.Info("failed to get versioned TykOasApiDefinition CR", "err", err.Error())
+			}
+
+			return err
+		}
+
+		if tykBaseOASCrd.Spec.Versioning.Default == v.Name {
+			versionOasAPI.Status.IsDefaultVersion = true
+		}
+
+		versionOasAPI.Status.BaseVersionName = tykBaseOASCrd.Name
+		versionOasAPI.Status.BaseVersionNamespace = tykBaseOASCrd.Namespace
+		versionOasAPI.Status.IsVersionedAPI = true
+
+		version := TykOasApiDefinitionVersion{
+			ID:   versionOasAPI.Status.ID,
+			Name: v.Name,
+		}
+
+		versions = append(versions, &version)
+
+		if err := r.Status().Update(ctx, &versionOasAPI); err != nil {
+			return err
+		}
+	}
+
+	versioning := TykOasApiDefinitionVersioning{
+		Default:  tykBaseOASCrd.Spec.Versioning.Default,
+		Enabled:  tykBaseOASCrd.Spec.Versioning.Enabled,
+		Key:      tykBaseOASCrd.Spec.Versioning.Key,
+		Location: tykBaseOASCrd.Spec.Versioning.Location,
+		Name:     tykBaseOASCrd.Spec.Versioning.Name,
+		Versions: versions,
+	}
+
+	byteVersioning, err := json.Marshal(versioning)
+	if err != nil {
+		return err
+	}
+
+	byteData := []byte(tykBaseOASDoc)
+
+	if byteData, err = jsonparser.Set(byteData, byteVersioning, "x-tyk-api-gateway", "info", "versioning"); err != nil {
+		return err
+	}
+
+	if err := klient.Universal.TykOAS().Update(ctx, baseAPIID, string(byteData)); err != nil {
```

Comment:
Instead of calling Update function here, just update the `tykBaseOASDoc` and let `createOrUpdateTykOASAPI` call update/create method.
This will reduce number of calls we are making to Tyk.